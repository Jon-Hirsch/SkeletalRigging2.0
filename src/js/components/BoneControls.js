import React from 'react';
import PropTypes from 'prop-types';
import dispatch from '../dispatch';

export default function BoneControls({ boneId, bones }) {
  const currentBone = bones.find(bone => bone.id === boneId);
  const children = bones.filter(bone => bone.parent === boneId);
  return (
    <div className="skeletal-rigging-bone-controls">
      <div className="bone-controls">
        <div>
          name:{' '}
          <input
            className="bone-name-input"
            type="text"
            value={currentBone.name}
            onChange={event =>
              updateBone(currentBone.id, 'name', event.target.value)}
          />
          x:{' '}
          <input
            className="bone-x-input"
            type="text"
            value={currentBone.x}
            onChange={event =>
              updateBone(currentBone.id, 'x', event.target.value)}
          />
          y:{' '}
          <input
            className="bone-y-input"
            type="text"
            value={currentBone.y}
            onChange={event =>
              updateBone(currentBone.id, 'y', event.target.value)}
          />
          length:{' '}
          <input
            className="bone-length-input"
            type="text"
            value={currentBone.boneLength}
            onChange={event =>
              updateBone(currentBone.id, 'boneLength', event.target.value)}
          />
          angle:{' '}
          <input
            type="text"
            className="bone-angle-input"
            value={currentBone.angle}
            onChange={event =>
              updateBone(currentBone.id, 'angle', event.target.value)}
          />
          <button
            className="add-bone-button"
            onClick={() => addBone(currentBone.id)}
          >
            Add Bone
          </button>
          {currentBone.parent && (
            <button
              className="remove-bone-button"
              onClick={() => removeBone(currentBone.id)}
            >
              Remove Bone
            </button>
          )}
        </div>
        <div />
      </div>
      {children.map(childBone => (
        <BoneControls key={childBone.id} boneId={childBone.id} bones={bones} />
      ))}
    </div>
  );
}

function addBone(id) {
  dispatch({
    type: 'ADD_BONE',
    parentId: id
  });
}

function removeBone(id) {
  dispatch({
    type: 'REMOVE_BONE',
    id
  });
}

function updateBone(id, field, value) {
  dispatch({
    type: 'UPDATE_BONE',
    id,
    field,
    value
  });
}

BoneControls.propTypes = {
  boneId: PropTypes.string.isRequired,
  bones: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      parent: PropTypes.string,
      x: PropTypes.string.isRequired,
      y: PropTypes.string.isRequired,
      boneLength: PropTypes.string.isRequired,
      angle: PropTypes.string.isRequired
    })
  ).isRequired
};
