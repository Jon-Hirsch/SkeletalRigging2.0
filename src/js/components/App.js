import React from 'react';
import PropTypes from 'prop-types';
import BoneControls from './BoneControls';

export default function App({ state }) {
  const baseBone = state.bones.find(bone => bone.parent === null);
  return <BoneControls boneId={baseBone.id} bones={state.bones} />;
}

App.propTypes = {
  state: PropTypes.shape({
    bones: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        parent: PropTypes.string,
        x: PropTypes.string.isRequired,
        y: PropTypes.string.isRequired,
        boneLength: PropTypes.string.isRequired,
        angle: PropTypes.string.isRequired
      })
    ).isRequired
  }).isRequired
};
