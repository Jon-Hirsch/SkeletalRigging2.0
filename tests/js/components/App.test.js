import React from 'react';
import App from '../../../src/js/components/App';
import BoneControls from '../../../src/js/components/BoneControls';

describe('App', () => {
  let shallowWrapper, state;

  function render() {
    return shallow(<App state={state} />);
  }

  beforeEach(() => {
    state = {
      bones: [
        {
          id: 'blah',
          name: 'anything',
          parent: 'body',
          x: '300',
          y: '320',
          boneLength: '200',
          angle: '270'
        },
        {
          id: 'body',
          name: 'body',
          parent: null,
          x: '300',
          y: '320',
          boneLength: '200',
          angle: '270'
        }
      ]
    };
    shallowWrapper = render();
  });

  test('should match snapshot', () => {
    expect(shallowWrapper.getElement()).toMatchSnapshot();
  });

  test('should render a boneControl with the base bone', () => {
    const boneControls = shallowWrapper.find(BoneControls);
    expect(boneControls.props().boneId).toEqual('body');
  });
});
