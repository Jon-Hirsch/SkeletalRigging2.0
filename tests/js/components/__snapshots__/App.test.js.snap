// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<BoneControls
  boneId="body"
  bones={
    Array [
      Object {
        "angle": "270",
        "boneLength": "200",
        "id": "blah",
        "name": "anything",
        "parent": "body",
        "x": "300",
        "y": "320",
      },
      Object {
        "angle": "270",
        "boneLength": "200",
        "id": "body",
        "name": "body",
        "parent": null,
        "x": "300",
        "y": "320",
      },
    ]
  }
/>
`;
